### PR TITLE
Update _goldens_io.dart to generate failure images during a size mism…

### LINKED ISSUE #141488: Golden tests do not write out failure images on a size mismatch
### Steps to reproduce

When using `matchesGoldenFile` / `LocalComparisonOutput` it will output failure images when the test image does not match the master image. However, this only occurs when the test/master images are identical sizes.  If they are mismatched sizes no failure images are written out. This can be frustrating, as we had a test image that was clearly twice as large as it should have been, but it wasn't easy to see the failure case.

1. Write some code like so:
```dart
await expectLater(
  find.text('Save'),
  matchesGoldenFile('save.png'),
);
```

2. Change the size of the 'Save' widget, or maybe easier change 'save.png' to be a different size.
3. Expect a error like this: 
> The following assertion was thrown while running async test code:
> Golden "goldens/save.png": Pixel test failed, image sizes do not match.
> Master Image: 200 X 100
> Test Image: 400 X 200

5. Look in the `failures` directory for `save_masterImage.png`, `save_testImage.png`, etc, and be surprised they don't exist.

### Expected results

Expected `failures/save_masterImage.png`, `failures/save_testImage.png` to exist.

### Actual results

They don't exist.

### Code sample

<details open><summary>Code sample</summary>

```dart
[Paste your code here]
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
❯ flutter doctor -v
[✓] Flutter (Channel stable, 3.16.8, on macOS 14.2.1 23C71 darwin-arm64, locale en-GB)
    • Flutter version 3.16.8 on channel stable at /opt/homebrew/Caskroom/flutter/3.16.0/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 67457e669f (8 days ago), 2024-01-16 16:22:29 -0800
    • Engine revision 6e2ea58a5c
    • Dart version 3.2.5
    • DevTools version 2.28.5

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/bramp/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15C500b
    • CocoaPods version 1.14.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] VS Code (version 1.85.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.81.20231227

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 14.2.1 23C71 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 120.0.6099.234

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by bramp**
Update the `matchesGoldenFile()` / `LocalComparisonOutput` code to generate failure images for golden tests that fail when the image sizes do not match. This can make it far quicker to identify what is wrong with the test image.

Fixes https://github.com/flutter/flutter/issues/141488

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ x I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] N/A I updated/added relevant documentation (doc comments with `///`).
- [x] N/A I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_test/lib/src/_goldens_io.dart
Comment (Piinks) [inline packages/flutter_test/lib/src/_goldens_io.dart:215]: This should be restored, we'll still want to make sure these are disposed of properly.

**Review by Piinks — COMMENTED** [review]
Hi @bramp welcome! Thanks for contributing. :)

This change will require a test that validates it works as expected.

File: packages/flutter_test/lib/src/_goldens_io.dart
Comment (bramp) [reply packages/flutter_test/lib/src/_goldens_io.dart:215]: These are disposed over here: https://github.com/flutter/flutter/blob/master/packages/flutter_test/lib/src/goldeComparisonResultns.dart#L397

A similar pattern is in https://github.com/flutter/flutter/blob/master/packages/flutter_test/lib/src/_goldens_io.dart#L265-L269 where the ComparisonResult is returned (with the included images) and not disposed there.

**Comment by bramp** [comment]
Ok tests added. Thanks for the review.

File: packages/flutter_test/lib/src/_goldens_io.dart
Comment (Piinks) [reply packages/flutter_test/lib/src/_goldens_io.dart:215]: > A similar pattern is in https://github.com/flutter/flutter/blob/master/packages/flutter_test/lib/src/_goldens_io.dart#L265-L269 where the ComparisonResult is returned (with the included images) and not disposed there.

They are disposed there. :)

https://github.com/flutter/flutter/blob/9c1665355d9b0deeec1be60b9e9137037defa877/packages/flutter_test/lib/src/_goldens_io.dart#L265-L274

> These are disposed over here: https://github.com/flutter/flutter/blob/master/packages/flutter_test/lib/src/goldeComparisonResultns.dart#L397

There appears to be a typo in this link, but I think you mean here:

https://github.com/flutter/flutter/blob/9c1665355d9b0deeec1be60b9e9137037defa877/packages/flutter_test/lib/src/goldens.dart#L390-L398

Which is correct! So in that case can you remove the other calls to dispose in `_goldens_io.dart`?

File: packages/flutter_test/test/goldens_test.dart
Comment (Piinks) [inline packages/flutter_test/test/goldens_test.dart:247]: Nice test! 👍

File: packages/flutter_test/lib/src/_goldens_io.dart
Comment (bramp) [reply packages/flutter_test/lib/src/_goldens_io.dart:215]: Sorry, not sure why that link 404s, but yes, you found the right place.

I'd be happy to make other changes, but I'm not sure about removing the other disposes. The images are only disposed if they aren't returned in the `diffs`. The only other `dispose` would be these https://github.com/flutter/flutter/blob/9c1665355d9b0deeec1be60b9e9137037defa877/packages/flutter_test/lib/src/_goldens_io.dart#L273-L274 and in that case, if those lines are reached the `ComparisonResult` will not contain the `masterImage`/`testImage` in their diff.

I think everything is disposed exactly once.

File: packages/flutter_test/test/goldens_test.dart
Comment (bramp) [reply packages/flutter_test/test/goldens_test.dart:247]: thanks, mostly copy and paste :)

**Comment by bramp** [comment]
Friendly ping, I think I addressed all comments.

File: packages/flutter_test/lib/src/_goldens_io.dart
Comment (Piinks) [reply packages/flutter_test/lib/src/_goldens_io.dart:215]: Ah I see, thanks for digging into the code with me on this. :)

**Review by Piinks — APPROVED** [review]
Flutter_LGTM
Thanks for contributing! This will need a second review, I've recruited someone for a second stamp!

**Review by justinmc — APPROVED** [review]
LGTM 👍 easy enough fix.

**Pull request merged**
